File: agora-easycall-ios-oc/AgoraSample/Login/AGSSelectMediaViewController.m
//
//  AGSEnterRoomViewController.m
//  AgoraSample
//
//  Created by apple on 15/9/15.
//  Copyright (c) 2015年 Agora. All rights reserved.
//

#import "AGSSelectMediaViewController.h"
#import "AGSCallBackView.h"
#import "AGSAgoraKitManager.h"
#import "AGSChatViewController.h"
#import "AGSEchoTestView.h"
#import "UIColor+Helper.h"

#define kRecordAudioFile @"textRecord.caf"

@interface AGSSelectMediaViewController ()

@property (weak, nonatomic) IBOutlet UITextField *roomNumberTextField;
@property (weak, nonatomic) IBOutlet UIButton *echoTest;
@property (weak, nonatomic) IBOutlet UILabel *videoLabel;
@property (weak, nonatomic) IBOutlet UILabel *voiceLabel;

@property (nonatomic, strong) AGSAlertView *alertView;
@property (nonatomic, strong) AGSEchoTestView *echoTestView;

@property (nonatomic, strong) AVAudioSession *audioSession;
@property (nonatomic, strong) AVAudioRecorder *audioRecorder;
@property (nonatomic, strong) AVAudioPlayer *audioPlayer;

@property (nonatomic, readwrite) BOOL isTesting;
@property (nonatomic, strong) NSTimer *recorderTimer;
@end

@implementation AGSSelectMediaViewController

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view.
}

- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}

- (void)viewWillAppear:(BOOL)animated
{
    [super viewWillAppear:animated];
    self.navigationController.navigationBarHidden = NO;
    
    //
    self.title = NSLocalizedString(@"Agora Easy Call", @"");
    [self.echoTest setTitle:NSLocalizedString(@"Test", @"") forState:UIControlStateNormal];
    self.roomNumberTextField.placeholder = NSLocalizedString(@"Enter Room No.", @"");
    self.videoLabel.text = NSLocalizedString(@"Join Video Call", @"");
    self.voiceLabel.text = NSLocalizedString(@"Join Voice Call", @"");
    
#ifdef DEBUG
        self.roomNumberTextField.text = @"1024";
#endif
}

- (void)viewWillDisappear:(BOOL)animated
{
    [super viewWillDisappear:animated];
}

#pragma mark - Actions

- (IBAction)didClickShowEchoTestView:(id)sender
{
    [self startTesting];
}

- (IBAction)didClickMenuButton:(UIButton *)sender
{
    [self.frostedViewController presentMenuViewController];
}

- (IBAction)didClickChatButton:(UIButton *)sender
{
    if ([self isValidInput]) {
        [self performSegueWithIdentifier:AGSKeySegueIdentifierChat sender:sender];
    }
}

- (BOOL)isValidInput
{
    [self.view endEditing:YES];
    if (self.roomNumberTextField.text.length && self.roomNumberTextField.text.length) {
        return YES;
    } else {
        UIAlertView *alertView = [[UIAlertView alloc] initWithTitle:nil
                                                            message:NSLocalizedString(@"Room No. can not be empty", @"")
                                                           delegate:nil
                                                  cancelButtonTitle:nil
                                                  otherButtonTitles:NSLocalizedString(@"Done", @""), nil];
        [alertView show];
        return NO;
    }
}

#pragma mark - Start/Stop testing

- (void)startTesting
{
    [self.view endEditing:YES];
    self.isTesting = YES;
    
    // Init and show
    [self setAlertViewStat:0];
    [self.alertView showAlertView];
}

- (void)stopTesting
{
    [self.alertView hideAlertView];
    self.isTesting = NO;
    // Stop recording and playing
    [self.recorderTimer invalidate];
    self.recorderTimer = nil;
    
    [self.audioRecorder stop];
//    [self.audioRecorder deleteRecording];
    [self.audioPlayer stop];
    [self.audioSession setActive:NO error:nil];
}

#pragma mark - Alert View Delegate

- (void)agsAlertView:(AGSAlertView *)alertView clickedButtonAtIndex:(NSInteger)buttonIndex
{
    if (buttonIndex == 0) {
        [self stopTesting];
    } else {
        [self startRecording];
    }
}

- (void)startRecording
{
    [self setAlertViewStat:1];

    [self.audioSession setCategory:AVAudioSessionCategoryRecord error:nil];
    [self.audioRecorder record];
    // Volume
    [NSTimer scheduledTimerWithTimeInterval:0.1 target:self selector:@selector(audioRecordPowerChange:) userInfo:nil repeats:YES];
    // Stop recoring
    self.recorderTimer = [NSTimer scheduledTimerWithTimeInterval:5 target:self selector:@selector(stopRecording) userInfo:nil repeats:NO];
    
}

- (void)stopRecording
{
    [self.audioRecorder stop];
}

#pragma mark - Audio Recorder/Player Delegate
// Recorder
- (void)audioRecorderDidFinishRecording:(AVAudioRecorder *)recorder successfully:(BOOL)flag
{
    if (!self.isTesting) {
        return;
    }
    
    [self setAlertViewStat:4];
    dispatch_after(dispatch_time(DISPATCH_TIME_NOW, (int64_t)(.3 * NSEC_PER_SEC)), dispatch_get_main_queue(), ^{
        
        [self.audioSession setCategory:AVAudioSessionCategoryPlayback error:nil];
        [self.audioPlayer play];
        [NSTimer scheduledTimerWithTimeInterval:0.1 target:self selector:@selector(audioPlayerPowerChange:) userInfo:nil repeats:YES];
    });
}

- (void)audioRecorderEncodeErrorDidOccur:(AVAudioRecorder *)recorder error:(NSError *)error
{
    [self setAlertViewStat:3];
}

// Player
- (void)audioPlayerDidFinishPlaying:(AVAudioPlayer *)player successfully:(BOOL)flag
{
    [self setAlertViewStat:2];
}

- (void)audioPlayerDecodeErrorDidOccur:(AVAudioPlayer *)player error:(NSError *)error
{
    [self setAlertViewStat:3];
}

#pragma mark - Others

- (void)audioRecordPowerChange:(NSTimer *)timer
{
    if (self.audioRecorder.isRecording) {
        [self.audioRecorder updateMeters];
        float power = [self.audioRecorder averagePowerForChannel:0];
        CGFloat progress = (power+160)/160.0;
        [self.echoTestView.volumeView setVolumenValue:progress];
    } else {
        [timer invalidate];
        timer = nil;
    }
}

- (void)audioPlayerPowerChange:(NSTimer *)timer
{
    if (self.audioPlayer.isPlaying) {
        [self.audioPlayer updateMeters];
        float power = [self.audioPlayer averagePowerForChannel:0];
        CGFloat progress = (power+160)/160.0;
        [self.echoTestView.volumeView setVolumenValue:progress];
    } else {
        [timer invalidate];
        timer = nil;
    }
}

- (void)setAlertViewStat:(NSInteger)stat
{
    switch (stat) {
        case 0: // Initial
            [self.echoTestView.volumeView setVolumenValue:0];
            [self.alertView setTitle:NSLocalizedString(@"Echo Test", @"")];
            [self.alertView setSubTitle:NSLocalizedString(@"Click Test to Start", @"") color:[UIColor colorWithHexString:@"B7B7B7"]];
            [self.alertView setRightButtonText:NSLocalizedString(@"Echo Test", @"") color:[UIColor colorWithHexString:@"1677CB"]];
            [self.alertView setLeftButtonText:NSLocalizedString(@"Cancel", @"")];
            [self.alertView setButtonEnabled:YES];
            break;
        case 1: // Doing
            [self.alertView setButtonEnabled:NO];
            [self.alertView setSubTitle:NSLocalizedString(@"Please speak to the phone", @"") color:[UIColor colorWithHexString:@"B7B7B7"]];
            [self.alertView setRightButtonText:NSLocalizedString(@"In Testing", @"") color:[UIColor colorWithHexString:@"999999"]];
            [self.alertView setLeftButtonText:NSLocalizedString(@"Cancel", @"")];
            break;
        case 2: // Success
            [self.alertView setButtonEnabled:YES];
            [self.alertView setSubTitle:NSLocalizedString(@"Click Test to Start", @"") color:[UIColor colorWithHexString:@"B7B7B7"]];
            [self.alertView setRightButtonText:NSLocalizedString(@"Test Again", @"") color:[UIColor colorWithHexString:@"1677CB"]];
            [self.alertView setLeftButtonText:NSLocalizedString(@"Done", @"")];
            break;
        case 3: // Error
            [self.alertView setButtonEnabled:YES];
            [self.alertView setSubTitle:NSLocalizedString(@"Click Test to Start", @"") color:[UIColor colorWithHexString:@"DCB059"]];
            [self.alertView setRightButtonText:NSLocalizedString(@"Test Again", @"") color:[UIColor colorWithHexString:@"1677CB"]];
            break;
        case 4: // Playing
            [self.alertView setSubTitle:NSLocalizedString(@"Detecting", @"")];
            break;
        default:
            break;
    }
}

#pragma mark -
- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender
{
    if (![segue.identifier isEqualToString:AGSKeySegueIdentifierChat]) {return;}
    
    AGSChatViewController *chatViewController = segue.destinationViewController;
    chatViewController.dictionary = @{AGSKeyChannel: self.roomNumberTextField.text};
    
    UIButton *button = (UIButton *)sender;
    if (button.tag == 0) {
        chatViewController.chatType = AGSChatTypeVideo;
    } else {
        chatViewController.chatType = AGSChatTypeAudio;
    }
}

#pragma mark - Testing Config

- (AGSAlertView *)alertView
{
    if (!_alertView) {
        _alertView = [[AGSAlertView alloc] init];
        _alertView.delegate = self;
        [_alertView addStatusView:self.echoTestView];
    }
    return _alertView;
}

- (AGSEchoTestView *)echoTestView
{
    if (!_echoTestView) {
        _echoTestView = [[[NSBundle mainBundle] loadNibNamed:@"AGSEchoTestView" owner:self options:nil] lastObject];
        [_echoTestView.volumeView setNumberOfBars:21];
        [_echoTestView.volumeView setBarsColorMin:[UIColor greenColor]];
        [_echoTestView.volumeView setBarsColorMax:[UIColor grayColor]];
        [_echoTestView.volumeView setVolumenValue:0];
    }
    return _echoTestView;
}

- (AVAudioSession *)audioSession
{
    if (!_audioSession) {
        _audioSession = [AVAudioSession sharedInstance];
        [_audioSession setCategory:AVAudioSessionCategoryPlayAndRecord error:nil];
        [_audioSession setActive:YES error:nil];
    }
    return _audioSession;
}

- (AVAudioRecorder *)audioRecorder
{
    if (!_audioRecorder) {
        NSDictionary *settings = @{AVFormatIDKey: @(kAudioFormatLinearPCM),
                                     AVSampleRateKey: @(8000),
                                     AVNumberOfChannelsKey: @(1),
                                     AVLinearPCMBitDepthKey: @(8),
                                     AVLinearPCMIsFloatKey: @(YES),
                                     AVLinearPCMIsBigEndianKey: @(NO)};
        NSError *error = nil;
        _audioRecorder = [[AVAudioRecorder alloc] initWithURL:[self filePath] settings:settings error:&error];
        _audioRecorder.delegate = self;
        _audioRecorder.meteringEnabled = YES;
    }
    
    return _audioRecorder;
}

- (AVAudioPlayer *)audioPlayer
{
    if (!_audioPlayer) {
        NSError *error = nil;
        _audioPlayer = [[AVAudioPlayer alloc]initWithContentsOfURL:[self filePath] error:&error];
        _audioPlayer.delegate = self;
        _audioPlayer.numberOfLoops = 0;
        _audioPlayer.meteringEnabled = YES;
        [_audioPlayer prepareToPlay];
    }
    return _audioPlayer;
}

- (NSURL *)filePath
{
    NSString *urlStr = [NSSearchPathForDirectoriesInDomains(NSDocumentDirectory, NSUserDomainMask, YES) lastObject];
    NSURL *url = [NSURL fileURLWithPath:[urlStr stringByAppendingPathComponent:kRecordAudioFile]];
    return url;
}

@end

File: agora-easycall-ios-swift/SampleSwift/AGSSelectMediaViewController.swift
//
//  AGSSelectMediaViewController.swift
//  SampleSwift
//
//  Created by let on 10/28/15.
//  Copyright © 2015 Agora. All rights reserved.
//

public struct AGSKeys {
    static let channel = "Channel";
    static let vendorKey = "VendorKey";
    static let username = "Username";
}

public struct AGSSegueIdentifiers {
    static let select = "Select";
    static let chat = "Chat";
}

import UIKit
import AVFoundation

class AGSSelectMediaViewController: AGSBaseViewController, AGSAlertViewDelegate, AVAudioRecorderDelegate, AVAudioPlayerDelegate {
    
    @IBOutlet weak var roomNumberTextField: UITextField!
    @IBOutlet weak var echoTest: UIButton!
    @IBOutlet weak var videoLabel: UILabel!
    @IBOutlet weak var voiceLabel: UILabel!
    
    //
    lazy var alertView: AGSAlertView = {
        let view = AGSAlertView()
        view.delegate = self
        view.addStatusView(self.echoTestView)
        return view
    }()
    lazy var echoTestView: AGSEchoTestView = {
        let view: AGSEchoTestView = NSBundle.mainBundle().loadNibNamed("AGSEchoTestView", owner: self, options: nil).last as! AGSEchoTestView
        view.volumeView?.numberOfBars = 21
        view.volumeView?.barsColorMin = UIColor.greenColor()
        view.volumeView?.barsColorMax = UIColor.grayColor()
        view.volumeView?.volumenValue = 0
        return view
    }()
    
    //
    lazy var audioSession: AVAudioSession = {
        let session = AVAudioSession.sharedInstance()
        do { try session.setCategory(AVAudioSessionCategoryPlayAndRecord) } catch {}
        do { try session.setActive(true) } catch {}
        return session
    }()
    lazy var audioRecorder: AVAudioRecorder = {
        var settings: [String : AnyObject] = [AVFormatIDKey: NSNumber(unsignedInt: kAudioFormatLinearPCM), AVSampleRateKey: NSNumber(int: 8000), AVNumberOfChannelsKey: NSNumber(int: 1), AVLinearPCMBitDepthKey: NSNumber(int: 8), AVLinearPCMIsFloatKey: NSNumber(bool: true), AVLinearPCMIsBigEndianKey: NSNumber(bool: false)]
        var recorder: AVAudioRecorder!
        do {
            recorder = try AVAudioRecorder(URL: self.filePath(), settings: settings)
            recorder.delegate = self
            recorder.meteringEnabled = true
            
        } catch {}
        return recorder
    }()
    lazy var audioPlayer: AVAudioPlayer = {
        var player = AVAudioPlayer()
        do {
            player = try AVAudioPlayer(contentsOfURL: self.filePath())
            player.delegate = self
            player.numberOfLoops = 0
            player.meteringEnabled = true
            player.prepareToPlay()
        } catch {}
        return player
    }()
    
    var testing: Bool!
    var recorderTimer = NSTimer()
    
    override func viewDidLoad() {
        super.viewDidLoad()

        // Do any additional setup after loading the view.
    }
    
    override func viewWillAppear(animated: Bool) {
        super.viewWillAppear(animated)
        
        navigationController?.navigationBarHidden = false
        
        //
        title = NSLocalizedString("Agora Easy Call", comment: "")
        echoTest.setTitle(NSLocalizedString("Test", comment: ""), forState: UIControlState.Normal)
        roomNumberTextField.placeholder = NSLocalizedString("Enter Room No.", comment: "")
        videoLabel.text = NSLocalizedString("Join Video Call", comment: "")
        voiceLabel.text = NSLocalizedString("Join Voice Call", comment: "")
        
//        if DEBUG == 1 {
//            roomNumberTextField.text = "1024"
//        }
    }
    
    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    // MARK: - Actions
    
    @IBAction func didClickShowEchoTestView(sender: AnyObject) {
        self.startTesting()
    }
    
    @IBAction func didClickMenuButton(sender: AnyObject) {
        frostedViewController.presentMenuViewController()
    }
    
    @IBAction func didClickChatButton(sender: AnyObject) {
        if isValidInput() {
            performSegueWithIdentifier(AGSSegueIdentifiers.chat, sender: sender)
        }
    }
    
    func isValidInput() -> Bool {
        view.endEditing(true)
        if roomNumberTextField.text?.characters.count > 0 {
            return true
        } else {
            UIAlertView(title: "", message: NSLocalizedString("Room No. can not be empty", comment: ""), delegate: nil, cancelButtonTitle: NSLocalizedString("Done", comment: "")).show()
            return false
        }
    }
    
    // MARK: - Start/Stop testing
    
    func startTesting() {
        view.endEditing(true)
        testing = true
        
        // Init and show
        setAlertViewStat(0)
        alertView.showAlertView()
    }
    
    func stopTesting() {
        alertView.hideAlertView()
        testing = false
        // Stop recording and playing
        recorderTimer.invalidate()
        
        audioRecorder.stop()
        audioPlayer.stop()
        do {
            try audioSession.setActive(false)
        } catch {
            
        }
        
    }
    
    // MARK: - Alert View Delegate
    func agsAlertView(alertView: AGSAlertView!, clickedButtonAtIndex buttonIndex: Int) {
        if buttonIndex == 0 {
            stopTesting()
        } else {
            startRecording()
        }
    }
    
    func startRecording() {
        setAlertViewStat(1)
        do { try audioSession.setCategory(AVAudioSessionCategoryRecord) } catch { }
        
        audioRecorder.record()
        // Volume
        NSTimer.scheduledTimerWithTimeInterval(0.1, target: self, selector: Selector("audioRecordPowerChange:"), userInfo: nil, repeats: true)
        // Stop recording
        recorderTimer = NSTimer.scheduledTimerWithTimeInterval(5, target: self, selector: Selector("stopRecording"), userInfo: nil, repeats: false)
    }
    
    func stopRecording() {
        audioRecorder.stop()
    }
    
    // MARK: - Audio Recorder/Player Delegate
    
    func audioRecorderDidFinishRecording(recorder: AVAudioRecorder, successfully flag: Bool) {
        if !testing {
            return;
        }
        
        setAlertViewStat(4)
        
        do { try audioSession.setCategory(AVAudioSessionCategoryPlayback) } catch {}
        audioPlayer.play()
        NSTimer.scheduledTimerWithTimeInterval(0.1, target: self, selector: Selector("audioPlayerPowerChange:"), userInfo: nil, repeats: true)
    }
    
    func audioRecorderEncodeErrorDidOccur(recorder: AVAudioRecorder, error: NSError?) {
        setAlertViewStat(3)
    }
    
    // Player
    
    func audioPlayerDidFinishPlaying(player: AVAudioPlayer, successfully flag: Bool) {
        setAlertViewStat(2)
    }
    
    func audioPlayerDecodeErrorDidOccur(player: AVAudioPlayer, error: NSError?) {
        setAlertViewStat(3)
    }
    
    // MARK: - Others
    
    func audioRecordPowerChange(timer: NSTimer) {
        if audioRecorder.recording {
            audioRecorder.updateMeters()
            let power = audioRecorder.averagePowerForChannel(0)
            let progress = (power + 160) / 160.0
            echoTestView.volumeView?.volumenValue = progress
        } else {
            timer.invalidate()
        }
    }
    
    func audioPlayerPowerChange(timer: NSTimer) {
        if audioPlayer.playing {
            audioPlayer.updateMeters()
            let power = audioPlayer.averagePowerForChannel(0)
            let progress = (power + 160) / 160.0
            echoTestView.volumeView?.volumenValue = progress
        } else {
            timer.invalidate()
        }
    }
    
    func setAlertViewStat(stat: Int) {
        switch (stat) {
        case 0:
            echoTestView.volumeView?.volumenValue = 0
            alertView.titleLabel?.text = NSLocalizedString("Echo Test", comment:"");
            alertView.setSubTitle(NSLocalizedString("Click Test to Start", comment:""), color: UIColor(hexString: "B7B7B7"))
            alertView.setRightButtonText(NSLocalizedString("Echo Test", comment:""), color: UIColor(hexString: "1677CB"))
            alertView.leftButton?.setTitle(NSLocalizedString("Cancel", comment:""), forState: UIControlState.Normal)
            alertView.setButtonEnabled(true)
            
        case 1:
            alertView.setButtonEnabled(false)
            alertView.setSubTitle(NSLocalizedString("Please speak to the phone", comment: ""), color: UIColor(hexString: "B7B7B7"))
            alertView.setRightButtonText(NSLocalizedString("In Testing", comment: ""), color: UIColor(hexString: "999999"))
            alertView.leftButton?.setTitle(NSLocalizedString("Cancel", comment: ""), forState: UIControlState.Normal)

            
        case 2:
            alertView.setButtonEnabled(true)
            alertView.setSubTitle(NSLocalizedString("Click Test to Start", comment:""), color: UIColor(hexString: "B7B7B7"))
            alertView.setRightButtonText(NSLocalizedString("Test Again", comment: ""), color: UIColor(hexString: "1677CB"))
            alertView.leftButton?.setTitle(NSLocalizedString("Done", comment: ""), forState: UIControlState.Normal)
            
        case 3:
            alertView.setButtonEnabled(true)
            alertView.setSubTitle(NSLocalizedString("Click Test to Start", comment:""), color: UIColor(hexString: "B7B7B7"))
            alertView.setRightButtonText(NSLocalizedString("Test Again", comment: ""), color: UIColor(hexString: "1677CB"))
            
        case 4:
            alertView.setSubTitle(NSLocalizedString("Detecting", comment:""), color: UIColor(hexString: "B7B7B7"))
        default:
            break
        }
        
    }
    
    // MARK: -
    
    override func prepareForSegue(segue: UIStoryboardSegue, sender: AnyObject?) {
        // Get the new view controller using segue.destinationViewController.
        // Pass the selected object to the new view controller.
        
        if segue.identifier == AGSSegueIdentifiers.chat {
            let chatViewController: AGSChatViewController = segue.destinationViewController as! AGSChatViewController
            chatViewController.channel(roomNumberTextField.text!)
            
            let button = sender as! UIButton
            if button.tag == 0 {
                chatViewController.type(AGSChatType.Video)
            } else {
                chatViewController.type(AGSChatType.Audio)
            }
        }
    }
    
    // MARK: - Testing Config
    func filePath() -> NSURL {
        var urlStr: String = NSSearchPathForDirectoriesInDomains(NSSearchPathDirectory.DocumentDirectory, NSSearchPathDomainMask.UserDomainMask, true).last!
        urlStr.appendContentsOf("/textRecord.caf")
        return NSURL(fileURLWithPath: urlStr)
    }

}
